### PR TITLE
Moving SyneRBI-Challenge nema-data utilities to SIRF

### PR DESCRIPTION
**Pull request by evgueni-ovtchinnikov**
## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added docstrings/doxygen in line with the guidance in the developer guide
- [ ] I have implemented unit tests that cover any new or modified functionality
- [x] The code builds and runs on my machine
- [ ] CHANGES.md has been updated with any functionality change

## Contribution Notes

Please read and adhere to the [contribution guidelines](https://github.com/SyneRBI/SIRF/blob/master/CONTRIBUTING.md).

Please tick the following: 

 - [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in SIRF (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License.

### CODE REVIEW

**Comment by KrisThielemans** [comment]
I'm not 100% sure if this should be in C++ TBH. I think most people will never check the C++ code, while they might do for the Python side. However, of course we can have it in C++ and an equivalent in Python as illustration.

Regarding the "incompatible projdata", if you post an error, I can try to help.

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans 
The following line in `cSTIR/tests/test7.cpp` 
```
		asm_.normalise(*background_sptr);
```
gives me
```
WARNING: BinNormalisationFromProjData: incompatible projection data:

Norm projdata info:
ProjDataInfoCylindricalNoArcCorr := 
Scanner parameters:=
  Scanner type := Siemens mMR
  Number of rings                          := 64
  Number of detectors per ring             := 504
  Inner ring diameter (cm)                 := 65.6
  Average depth of interaction (cm)        := 0.7
  Distance between rings (cm)              := 0.40625
  Default bin size (cm)                    := 0.208626
  View offset (degrees)                    := 0
  Maximum number of non-arc-corrected bins := 344
  Default number of arc-corrected bins     := 344
  Energy resolution         := 0.145
  Reference energy (in keV) := 511
  Number of blocks per bucket in transaxial direction         := 1
  Number of blocks per bucket in axial direction              := 2
  Number of crystals per block in axial direction             := 8
  Number of crystals per block in transaxial direction        := 9
  Number of detector layers                                   := 1
  Number of crystals per singles unit in axial direction      := 16
  Number of crystals per singles unit in transaxial direction := 9
  Scanner geometry (BlocksOnCylindrical/Cylindrical/Generic)  := Cylindrical
End scanner parameters:=

start vertical bed position (mm) := 0
start horizontal bed position (mm) := 0

TOF mashing factor in data:      0
Number of TOF positions in data: 1

Segment_num range:           (-1, 1)
Number of Views:                126
Number of axial positions per seg: {115 127 115 }
Number of tangential positions: 344
Azimuthal angle increment (deg):   1.42857
Azimuthal angle extent (deg):      180
ring differences per segment: 
(-16,-6)(-5,5)(6,16)
End :=

Emission projdata info (made non-TOF if norm is non-TOF):

ProjDataInfoCylindricalNoArcCorr := 
Scanner parameters:=
  Scanner type := Siemens mMR
  Number of rings                          := 64
  Number of detectors per ring             := 504
  Inner ring diameter (cm)                 := 65.6
  Average depth of interaction (cm)        := 0.7
  Distance between rings (cm)              := 0.40625
  Default bin size (cm)                    := 0.208626
  View offset (degrees)                    := 0
  Maximum number of non-arc-corrected bins := 344
  Default number of arc-corrected bins     := 344
  Energy resolution         := 0.145
  Reference energy (in keV) := 511
  Number of blocks per bucket in transaxial direction         := 1
  Number of blocks per bucket in axial direction              := 2
  Number of crystals per block in axial direction             := 8
  Number of crystals per block in transaxial direction        := 9
  Number of detector layers                                   := 1
  Number of crystals per singles unit in axial direction      := 16
  Number of crystals per singles unit in transaxial direction := 9
  Scanner geometry (BlocksOnCylindrical/Cylindrical/Generic)  := Cylindrical
End scanner parameters:=

start vertical bed position (mm) := 0
start horizontal bed position (mm) := -19

TOF mashing factor in data:      0
Number of TOF positions in data: 1

Segment_num range:           (-1, 1)
Number of Views:                126
Number of axial positions per seg: {115 127 115 }
Number of tangential positions: 344
Azimuthal angle increment (deg):   1.42857
Azimuthal angle extent (deg):      180
ring differences per segment: 
(-16,-6)(-5,5)(6,16)
End :=

--- (end of incompatible projection data info)---
```
(note the difference in `start horizontal bed position (mm)`), and the rest of the code is not executed.

I tried different declarations of `background_sptr` to no avail.

**Comment by KrisThielemans** [comment]
Cutting out some stuff
```c++
std::string f_template = append_path(path, "mMR_template_span11_small.hs");
CREATE_OBJECT(STIRAcquisitionData, STIRAcquisitionDataInFile,
		acq_data_template, templ_sptr, f_template.c_str());
...
converter.sinograms_and_randoms_from_listmode
		(lm_data, 0, 10, acq_data_template, sinograms_sptr, randoms_sptr);
...
PETAttenuationModel::compute_ac_factors(templ_sptr, mu_map_sptr, acf_sptr, iacf_sptr);
```
The last line seems incorrect. the template should no longer be used, as that doesn't know about bed positions, time frames etc actually used. Instead, use `sinograms_sptr`

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans thanks a lot - seems to work ok now!

File: examples/Python/PET/listmode_to_sinograms.py [hunk 1]
```diff
@@ -77,6 +77,15 @@ def main():
 
     # create listmode-to-sinograms converter object
     lm2sino = pet.ListmodeToSinograms()
+    lm_data = pet.ListmodeData(list_file)
+    sino = pet.AcquisitionData(acq_data_template)
+    rand = pet.AcquisitionData(acq_data_template)
+    sino, rand = lm2sino.sinograms_and_randoms_from_listmode(lm_data, 0, 10, acq_data_template)
+    print(sino.shape)
+    print(sino.norm())
```
Comment (KrisThielemans) [inline examples/Python/PET/listmode_to_sinograms.py:85]: let's rename `sino` to `prompts`. I'd prefer to rename `rand` to `randoms_estimate` (or at least, say that in a comment). They are not the randoms of course.

Also rename the function to `get_prompts_and_randoms(...)`. However, I'm not a great fan of this I'm afraid. We have `set_input` etc, and here a (presumably static) function that takes an lm_data and time-frame info. I don't see a good reason to have both interfaces. So, I suppose I suggest to have no arguments for this function (i.e. bunch of sets, a set_up(), and then call this). It raises the question why we have it, aside from having a one-liner.

We're mixing 2 kinds of interfaces here, which creates confusion.

If you don't like the `set` functions. we could have a constructor function for Listmode2Sinograms
```python
def __init__(self, list_file, interval, template)
```

Same comments will apply on the C++ side of course.

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
not sure about the direction this is going I'm afraid. Some more detail in the comments.

File: examples/Python/PET/listmode_to_sinograms.py [hunk 2]
```diff
@@ -77,6 +77,15 @@ def main():
 
     # create listmode-to-sinograms converter object
     lm2sino = pet.ListmodeToSinograms()
+    lm_data = pet.ListmodeData(list_file)
+    sino = pet.AcquisitionData(acq_data_template)
+    rand = pet.AcquisitionData(acq_data_template)
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply examples/Python/PET/listmode_to_sinograms.py]: debugging leftovers - used this script for testing, now restored

File: src/iUtilities/include/sirf/iUtilities/DataHandle.h [hunk 3]
```diff
@@ -279,6 +279,22 @@ getObjectSptrFromHandle(const void* h, std::shared_ptr<Object>& sptr) {
 	sptr = *ptr_sptr;
 }
 
+template<class Object>
+void
+setHandleObjectSptr(void* h, std::shared_ptr<Object>& sptr) {
+	ObjectHandle<Object>* handle = (ObjectHandle<Object>*)h;
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/iUtilities/include/sirf/iUtilities/DataHandle.h]: thanks!

File: src/xSTIR/cSTIR/cstir.cpp [hunk 4]
```diff
@@ -628,6 +649,22 @@ void* cSTIR_createPETAttenuationModel(const void* ptr_img, const void* ptr_am)
 	CATCH;
 }
 
+extern "C"
+void* cSTIR_computeACF(const void* ptr_sino, const void* ptr_mumap, void* ptr_acf, void* ptr_iacf)
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/cstir.cpp]: My `acf_sptr` corresponds to `ac_factors` and `iacf_sptr` to `acf_factors` in your function `estimate_scatter` in `SyneRBI-Challenge/src/SIRF_data_preparation/data_preparation_setup.py`. I just shortened `ac_factors` to `acf` for brevity, and `*iacf_sptr` is just the inverse of `*acf_sptr` (I checked that `(*acf_sptr)*(*iacf_sptr)` is all ones). No problem to rename back.

File: src/xSTIR/cSTIR/tests/test7.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
+/*
+SyneRBI Synergistic Image Reconstruction Framework (SIRF)
+Copyright 2018 - 2020 Rutherford Appleton Laboratory STFC
+Copyright 2018 - 2020 University College London
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/tests/test7.cpp]: corrected

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 6]
```diff
@@ -534,52 +375,453 @@ The actual algorithm is described in
...
+        float get_time_at_which_num_prompts_exceeds_threshold(const unsigned long threshold) const;
+
+		void sinograms_and_randoms_from_listmode(
+			const STIRListmodeData& lm_data, double start, double stop,
+			const STIRAcquisitionData& acq_data_template,
+			std::shared_ptr<STIRAcquisitionData>& sinograms_sptr,
+			std::shared_ptr<STIRAcquisitionData>& randoms_sptr)
+		{
+			set_input(lm_data);
+			set_output("sinograms");
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
This PR introduces convenience functions for PET data preparation. In particular, it adds `sirf.STIR.AcquisitionSensitivityModel.compute_attenuation_factors`  and `sirf.STIR.ListmodeToSinograms.prompts_and_randoms_from_listmode` (and corresponding C/C++ functions).

For the challenge, we could have just added them in the Challenge repo, but it's obviously better to have them in SIRF itself.

However, I think the current PR has several problems.
- it adds "one-liner" to class interfaces, introducing duplication in functionality
- most of it is PET specific but they are at "top-level" sirf.STIR.
   - `sirf.STIR.AcquisitionSensitivityModel.compute_attenuation_factors` does not make sense for SPECT, where they attenuation image has to be entered explicitly in the `AcquisitionModel`. So, that function (if we keep it) needs to be moved to `PETAttenuationModel` (where it already is in C++).
   - randoms do not exist in SPECT

My preferred solution is to have `sirf.STIR.PET_data_preparation` Python module (it could be on the C++ side as well of course, but more work) where these kind of one-liners are located (as functions, not as class members). I'm not sure if that's easy to do quickly though.

Possibly best things to do in this PR is to:
- move  `sirf.STIR.AcquisitionSensitivityModel.compute_attenuation_factors`  to a PET specific model. If that doesn't exist on the Python side, I'm afraid I would *not* add this to the class.
- drop the list-mode one-liner

Aside from this, there are huge diffs in `stir_x.h`, probably just due to moving things around. I haven't reviewed any of that.

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 7]
```diff
@@ -534,52 +375,457 @@ The actual algorithm is described in
...
+	/*!
+	\ingroup PET
+	\brief Attenuation model.
+
+	*/
+
+	class PETAttenuationModel : public PETAcquisitionSensitivityModel {
+	public:
+		PETAttenuationModel(STIRImageData& id, PETAcquisitionModel& am);
+		// multiply by bin efficiencies
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 8]
```diff
@@ -534,52 +375,457 @@ The actual algorithm is described in
...
+		// divide by bin efficiencies
+		virtual void normalise(STIRAcquisitionData& ad) const;
+		static void compute_ac_factors(
+			std::shared_ptr<STIRAcquisitionData> acq_templ_sptr,
+			std::shared_ptr<STIRImageData> mu_map_sptr,
+			std::shared_ptr<PETAttenuationModel>& asm_sptr,
+			std::shared_ptr<STIRAcquisitionData>& acf_sptr,
+			std::shared_ptr<STIRAcquisitionData>& iacf_sptr)
+		{
+			PETAcquisitionModelUsingRayTracingMatrix acq_mod;
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 9]
```diff
@@ -534,52 +375,457 @@ The actual algorithm is described in
...
+			PETAttenuationModel acq_sens_mod = *asm_sptr;
+			acq_sens_mod.set_up(acq_templ_sptr->get_exam_info_sptr(),
+				acq_templ_sptr->get_proj_data_info_sptr()->create_shared_clone());
+			acq_mod.set_asm(asm_sptr);
+			std::cout << "applying attenuation (please wait, may take a while)...\n";
+			acq_sens_sptr = acq_templ_sptr->clone();
+			acq_sens_sptr->fill(1.0);
+			acq_sens_mod.unnormalise(*acq_sens_sptr);
+		}
+*/
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

File: examples/Python/PET/test_data_preparation.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
+import os
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply examples/Python/PET/test_data_preparation.py:1]: yes it is, will be removed after #1241 is merged

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 11]
```diff
@@ -534,52 +375,457 @@ The actual algorithm is described in
...
+	*/
+
+	class PETAttenuationModel : public PETAcquisitionSensitivityModel {
+	public:
+		PETAttenuationModel(STIRImageData& id, PETAcquisitionModel& am);
+		// multiply by bin efficiencies
+		virtual void unnormalise(STIRAcquisitionData& ad) const;
+		// divide by bin efficiencies
+		virtual void normalise(STIRAcquisitionData& ad) const;
+		static void compute_ac_factors(
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h:558]: done

File: src/xSTIR/pSTIR/STIR.py [hunk 12]
```diff
@@ -1723,6 +1735,18 @@ def invert(self, ad):
         check_status(fd.handle)
         return fd
 
+    @staticmethod
+    def compute_attenuation_factors(sinograms, mu_map):
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/pSTIR/STIR.py:1739]: done

File: src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h [hunk 13]
```diff
@@ -534,52 +375,457 @@ The actual algorithm is described in
...
+//			std::shared_ptr<PETAttenuationModel>
+//				asm_sptr(new PETAttenuationModel(*mu_map_sptr, acq_mod));
+			PETAttenuationModel& acq_sens_mod = *asm_sptr;
+			acq_sens_mod.set_up(acq_templ_sptr->get_exam_info_sptr(),
+				acq_templ_sptr->get_proj_data_info_sptr()->create_shared_clone());
+			acf_sptr = acq_templ_sptr->clone();
+			acf_sptr->fill(1.0);
+			iacf_sptr = acf_sptr->clone();
+			acq_sens_mod.unnormalise(*acf_sptr);
+			acq_sens_mod.normalise(*iacf_sptr);
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/xSTIR/cSTIR/include/sirf/STIR/stir_x.h]: done

**Comment by evgueni-ovtchinnikov** [comment]
> randoms don't exist in SPECT

So, current `ListmodeToSinograms` will not work for SPECT data? Does STIR now has a converter that works for SPECT?

**Comment by KrisThielemans** [comment]
Listmode 2 sinograms wil work for SPECT as long as you don't ask for the randoms.

**Comment by evgueni-ovtchinnikov** [comment]
> Listmode 2 sinograms wil work for SPECT as long as you don't ask for the randoms.

So far as I can see, `ListmodeToSinograms::set_up` assigns `randoms.sptr` to an empty `STIRAcquisitionData` object. Should we then add checking the size of `randoms` data in `estimate_randoms()`, `save_randoms()` and `get_randoms_sptr()` and issuing a warning if it is zero? In SIRF 4.0 we could do a more proper fix via inheritance from a common base class.

**Comment by KrisThielemans** [comment]
Can be an error I think, but yes, good plan

**Comment by KrisThielemans** [comment]
@evgueni-ovtchinnikov let's try to finish this off and merge. Given above problems with `Listmode2Sinograms::sinograms_and_randoms_from_listmode` I suggest to cut it out, and keep relevant lines in the python code. you could make it a function there obviously.

Then add something to `CHANGES.md` and go ahead.

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
I'm sorry, but although this last commit is fine, I don't like it. I find introducing multiple ways to the class interface to do  something in the end less productive. How are we going to handle the randoms now? Add some other member to attempt a one liner? Where do we stop?

We need to move on, and we keep circling. so please abandon this PR and just add equivalent functions in the Challenge PR. Sorry.

We could pick it up later, but it'd need design decisions.

- `compute_ac_factors` 
    - C++: it is a static function of `PETAttenuationModel`, that takes a `PETAttenuationModel` as argument, that's pretty weird.
    - Python: move `sirf.STIR.AcquisitionSensitivityModel.compute_attenuation_factors` to a `sirf.STIR.PETAttenuationModel`. However, that doesn't exist on the Python side yet, so I don't know how much work that'd be. This would be worth doing though, as really the current `AcquisitionSensitivityModel` constructor taking an `ImageData` as argument should sit in `PETAttenuationModel`. As that breaks backwards compatibility, it's for 4.0.
- remove `prompts_*_from_listmode` completely.
- remove `test_data_preparation.py`
- change `CHANGES.md` to say you've added a convenience function `compute_af_factors`
- make sure that the attenuation model is NOT returned from `compute_af_factors`
- rename `test7.cxx` to something that says what it's testing.

**Comment by KrisThielemans** [comment]
Right, I see this was merged. Please revert the merge. Sorry

**Pull request merged**